### PR TITLE
[🔨 WIP 🔨] Support running on ATS grid

### PR DESCRIPTION
**Pull request by nagpalkaran95**
https://browserstack.atlassian.net/browse/HST-722

### CODE REVIEW

File: bin/helpers/atsHelper.js [hunk 1]
```diff
@@ -0,0 +1,82 @@
+// TODO: confirm if we need to add command line args. if yes then decide priority
```
Comment (asambstack) [inline bin/helpers/atsHelper.js]: Refactoring required for TODO

**Review by asambstack — CHANGES_REQUESTED** [review]

File: bin/commands/runs.js [hunk 2]
```diff
@@ -144,6 +147,23 @@ module.exports = function run(args, rawArgs) {
...
+
+        if (gridDetails.isTrialGrid) {
+          bsConfig.connection_settings.local = true;
+          bsConfig.connection_settings.local_inferred = true;
+          bsConfig.connection_settings.local_mode = 'on-demand';
+        }
+
+        Constants.turboScaleObj.gridUrl = gridDetails.cypressUrl;
+        Constants.turboScaleObj.uploadUrl = gridDetails.cypressUrl + '/upload';
+        Constants.turboScaleObj.buildUrl = gridDetails.cypressUrl + '/build';
```
Comment (asambstack) [inline bin/commands/runs.js:163]: Won't this throw an `undefined` exception if the request for grid details fail in `getTurboScaleGridDetails` at https://github.com/browserstack/browserstack-cypress-cli/pull/736/files#diff-6ad671bcacc06082182a70064d2229f5537447360f11c88e1e2d92d800e9befeR59-R77

File: bin/helpers/atsHelper.js (hunk 1 above)
Comment (nagpalkaran95) [reply bin/helpers/atsHelper.js]: resolved TODO

File: bin/commands/runs.js (hunk 2 above)
Comment (nagpalkaran95) [reply bin/commands/runs.js]: added that handling.. re-review

**Review by asambstack — DISMISSED** [review]
LGTM

File: bin/helpers/utils.js [hunk 3]
```diff
@@ -1141,6 +1147,13 @@
   }
 
   logger.debug(`${files ? files.length : 0} spec files found`);
+
+  if (turboScaleSession) {
+    // remove unwanted path prefix for turboscale
+    files = files.map((x) => { return path.join(testFolderPath, x.split(testFolderPath)[1]) })
```
[the remark this answers is not in the document]
Comment (nagpalkaran95) [reply bin/helpers/utils.js:1157]: we are just trimming few path strings…removing prefix as we expect trimmed down path…
using path.join to be os agnostic.

[the remark this answers is not in the document]
Comment (nagpalkaran95) [reply bin/helpers/utils.js:1157]: we are just trimming few path strings…removing prefix as we expect trimmed down path…
using path.join to be os agnostic.
no impact

File: bin/helpers/atsHelper.js [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+      });
+    });
+  } catch (err) {
+    logger.error(`Failed to find TurboScale Grid: ${err}: ${err.stack}`);
+  }
+};
+
+exports.patchCypressConfigFileContent = (bsConfig) => {
+  try {
+    let cypressConfigFileData = fs.readFileSync(path.resolve(bsConfig.run_settings.cypress_config_file)).toString();
```
[the remark this answers is not in the document]
Comment (nagpalkaran95) [reply bin/helpers/atsHelper.js:87]: no impact can be ignored. check above comment

File: bin/helpers/atsHelper.js [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+    fs.writeFileSync(path.resolve(bsConfig.run_settings.patched_cypress_config_file), patchedConfigFileData);
+  } catch(e) {
+    logger.error(`Encountered an error when trying to patch ATS Cypress Config File ${e}`);
+    return {};
+  }
+}
+
+exports.atsFileCleanup = (bsConfig) => {
+  const filePath = path.resolve(bsConfig.run_settings.patched_cypress_config_file);
```
[the remark this answers is not in the document]
Comment (nagpalkaran95) [reply bin/helpers/atsHelper.js:114]: no impact can be ignored. check above comment

**Review by asambstack — APPROVED** [review]

**Comment by nagpalkaran95** [comment]
Semgrep issue can be ignored and force merged as per slack thread - https://browserstack.slack.com/archives/CR42BQ3AQ/p1702374508668739?thread_ts=1702357970.975999&cid=CR42BQ3AQ

**Review by KrishnaSuravarapu — APPROVED** [review]

**Pull request merged**